servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpMessageDiscardWatchdogClientFilter.java

```diff
@@ -0,0 +1,132 @@
+                    final AtomicReference<Publisher<?>> reference = request.context()
+                            .computeIfAbsent(MESSAGE_PUBLISHER_KEY, key -> new AtomicReference<>());
+                    assert reference != null;
+                    final Publisher<?> previous = reference.getAndSet(response.messageBody());
+                    if (previous != null) {
+                        // If a previous message exists, the Single<StreamingHttpResponse> got resubscribed to
+                        // (i.e. during a retry) and so previous message body needs to be cleaned up.
+                        LOGGER.warn("Automatically draining previous HTTP response message body that was " +
+                                "not consumed. Users-defined retry logic must drain response payload before " +
+                                "retrying.");
+
+                        previous.ignoreElements().subscribe();
+                    }
+
+                    return response.transformMessageBody(msgPublisher -> msgPublisher.beforeSubscriber(() -> {
+                        final AtomicReference<?> maybePublisher = request.context().get(MESSAGE_PUBLISHER_KEY);
+                        if (maybePublisher != null) {
+                            maybePublisher.set(null);
+                        }
+                        return HttpMessageDiscardWatchdogServiceFilter.NoopSubscriber.INSTANCE;
+                    }));
+                });
+            }
+        };
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return HttpExecutionStrategies.offloadNone();
+    }
+
+    private static final class CleanerStreamingHttpClientFilterFactory implements StreamingHttpClientFilterFactory {
+        @Override
+        public StreamingHttpClientFilter create(final FilterableStreamingHttpClient client) {
+            return new StreamingHttpClientFilter(client) {
+                @Override
+                protected Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
+                                                                final StreamingHttpRequest request) {
+                    return delegate.request(request).onErrorResume(originalThrowable -> Single.defer(() -> {
+                        final AtomicReference<?> maybePublisher = request.context().get(MESSAGE_PUBLISHER_KEY);
+                        if (maybePublisher != null) {
+                            Publisher<?> message = (Publisher<?>) maybePublisher.get();
+                            if (message != null) {
+                                // No-one subscribed to the message (or there is none), so if there is a message
+                                // proactively clean it up.
+                                return message.ignoreElements().concat(Single.failed(originalThrowable));
+                            }
+                        }
+                        return Single.failed(originalThrowable);
```

Comment:
For both `return` cases we also need to apply `.shareContextOnSubscribe()` at the end